main/src/ca/uwaterloo/flix/language/phase/jvm/GenClosureClasses.scala

```diff
@@ -124,7 +124,7 @@ object GenClosureClasses {
     invokeMethod.visitCode()
 
     // Free variables
-    val frees = defn.formals.take(freeVars.length).map(x => FreeVar(x.sym, x.tpe))
+    val frees = defn.formals.take(freeVars.length).map(x => (x.sym, x.tpe))
```

Comment:
@magnus-madsen could this be the problem? Maybe we need to do something about the free variables that have been inlined?

Reply:
I would guess that if you are inlining free vars then something has to change, yes.

Reply:
@ChrBondi Can you post an example program and describe how it is transformed by your inliner?

Reply:
`def main(_args: Array[String]): Int32 & Impure =
`
`
 let x = 5;
`
`
 let y = {print(""); 3};
`
`
 let l = (z -> x+y+z);
`
`
 l(8)
`
Since x is pure we want to inline it. The inliner runs after ClosureConv and LambdaLift
`
def main(_args: Array[String]): Int32 & Impure =
`
`
 let y = {print(""); 3};
`
`
 Closure(main2, [5, y ])*(8i32, )
`


`
def main2(x: Int32, y: Int32, z: Int32): Int32 = add(add(x, y), z)
`
The expressions the the brackets of the closure are the freevars which have now been turned into expressions and where x has been inlined into

Reply:
> @JonathanStarup Maybe you can help? I am not 100% sure.
> 
> At least, can we get a description of what this code does.

This code takes the first `freeVars.length` formals and maps them to fields `clo$i` and the rest of the formals are mapped to `arg$i`. So if the formals list and the length of freeVars does not align there would be an error here

Reply:
> > @JonathanStarup Maybe you can help? I am not 100% sure.
> > At least, can we get a description of what this code does.
> 
> This code takes the first `freeVars.length` formals and maps them to fields `clo$i` and the rest of the formals are mapped to `arg$i`. So if the formals list and the length of freeVars does not align there would be an error here

The lenghts of freevars should not change so i don't think thats a problem

Running Test.Array.Flix in LibarySuite throws the NullPointer i mentioned earlier but i can't see which testcase is throwing it

Reply:
alright nice, ill check it out

Reply:
Heres a very  small program with the null pointer. In the jvm code there are more closure arguments than in master but that is maybe to be expected?
```
def main(_args: Array[String]): Int32 & Impure =
    let a = Array.filter(_ -> false, [2]);
    a.length
```

Reply:
I can't see anything wrong with the backend stuff or var numbering... im not sure what's wrong 🤔

Reply:
Its seems like there is a problem with local defs and your inlining. something about giving a closure itself as a closure argument. Here is the generated code showing the null introduction.
```java
Clo%filter%823965 var6 = null;
Clo%filter%823965 var11 = new Clo%filter%823965();
var11.clo0 = var4;
var11.clo1 = (Object)var2;
var11.clo2 = (Fn1%Int32%Bool)var1;
var11.clo3 = (Object)var5;
var11.clo4 = (Fn1%Int32%Obj)var6;
var11.clo5 = (Fn1%Int32%Obj)var6;
var11.clo4 = var11;
var6 = var11;
```

Reply:
Okay, in summary: The self-reference(s) need special treatment in the backend I've added the ability to have multiple self-references in [this PR](https://github.com/JonathanStarup/flix/commit/33d9f1e6e4f16ab9a3d1ec50183774141f6eb28e) but 1) it doesn't fix the problem because it comes before your phase and 2) maybe there shouldn't be multiple self-references?

I see two solutions:
1) In your code, be careful not to introduce new self-references in the captured variables of the closures and not to move the existing one.
2) We can merge the LetRec ast node with the relevant closure ast node. This would remove the special treatment of self-references

This is maybe a decision for you @magnus-madsen?

Reply:
@JonathanStarup Does the problem go away if let-recs are left alone? Or is it unrelated to letrec?

Reply:
Is the problem that when you inline into recursive functions then you have to be careful with where they call themselves?

Reply:
It sounds like we may need to meet and discuss this together.

Reply:
> @JonathanStarup Does the problem go away if let-recs are left alone? Or is it unrelated to letrec?

I would say so yes. More precisely its the arguments of the closure related to it that should not be changed

Reply:
> It sounds like we may need to meet and discuss this together.

Yeah I think that sounds good